src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/CastHelpers.cs

```diff
@@ -492,5 +497,120 @@ private static unsafe void ArrayTypeCheck_Helper(object obj, void* elementType)
+        private static bool AreTypesEquivalent(MethodTable* pMTa, MethodTable* pMTb)
+        {
+            if (pMTa == pMTb)
+            {
+                return true;
+            }
+
+            if (!pMTa->HasTypeEquivalence || !pMTb->HasTypeEquivalence)
+            {
+                return false;
+            }
+
+            return RuntimeHelpers.AreTypesEquivalent(pMTa, pMTb);
+        }
+
+        [DebuggerHidden]
+        private static bool IsNullableForType(MethodTable* typeMT, MethodTable* boxedMT)
+        {
+            if (!typeMT->IsNullable)
+            {
+                return false;
+            }
+
+            MethodTable *pMTNullableArg = typeMT->InstantiationArg0();
+            if (pMTNullableArg == boxedMT)
+            {
+                return true;
+            }
+            else
+            {
+                return AreTypesEquivalent(pMTNullableArg, boxedMT);
+            }
+        }
+
+        [DebuggerHidden]
+        internal static void Unbox_Nullable(ref byte destPtr, MethodTable* typeMT, object? obj)
+        {
+            if (obj == null)
+            {
+                InitValueClass(ref destPtr, typeMT);
+            }
+            else
+            {
+                if (!IsNullableForType(typeMT, RuntimeHelpers.GetMethodTable(obj)))
+                {
+                    // For safety's sake, also allow true nullables to be unboxed normally.
+                    // This should not happen normally, but we want to be robust
+                    if (typeMT == RuntimeHelpers.GetMethodTable(obj))
+                    {
+                        Unsafe.CopyBlockUnaligned(ref destPtr, ref RuntimeHelpers.GetRawData(obj), typeMT->GetNumInstanceFieldBytes());
+                        return;
+                    }
+                    CastHelpers.ThrowInvalidCastException(obj, typeMT);
+                }
+
+                // Set the hasValue field on the Nullable type. It MUST always be placed at the start of the object.
+                Unsafe.As<byte, bool>(ref destPtr) = true;
+                ref byte destValuePtr = ref typeMT->GetNullableValueFieldReferenceAndSize(ref destPtr, out uint size);
+                Unsafe.CopyBlockUnaligned(ref destValuePtr, ref RuntimeHelpers.GetRawData(obj), size);
+            }
+        }
+
+        [DebuggerHidden]
+        internal static ref byte Unbox_Helper(MethodTable* pMT1, object obj)
+        {
+            // must be a value type
+            Debug.Assert(pMT1->IsValueType);
+
+            MethodTable* pMT2 = RuntimeHelpers.GetMethodTable(obj);
+            if ((pMT1->IsPrimitive && pMT2->IsPrimitive &&
+                pMT1->GetPrimitiveCorElementType() == pMT2->GetPrimitiveCorElementType()) ||
+                AreTypesEquivalent(pMT1, pMT2))
+            {
+                return ref RuntimeHelpers.GetRawData(obj);
+            }
+
+            CastHelpers.ThrowInvalidCastException(obj, pMT1);
+            return ref Unsafe.AsRef<byte>(null);
```

Comment:
Invert the if here to remove the unreachable return?

Reply:
To do this, I expect that `ThrowInvalidCastException` would have to `throw` in C# to make the JIT realize that it is cold code and produce the desired code layout.

Reply:
Ah, only now I've noticed that it does a QCall instead of direct throw. What's the point of the runtime call here?

Reply:
Yes. And there are resource file changes that I would need to do to get the right string. I'll do it if you want @jkotas, but honestly this isn't so bad as it is. I couldn't find any impact on performance from the generated assembly code.

Reply:
I agree that rewriting cast exception message formatting in C# is a problem for another day.

I think you can add unreachable `throw` to `ThrowInvalidCastException` to allow JIT to reorder the basic blocks for you instead of doing it manually (left suggestion what it may look like).